### PR TITLE
Implement per channel logginig again

### LINKED ISSUE #282: Implement logging loss per-chanel
### Is your feature request related to a problem? Please describe.

Currently there's no way to monitor how individual chanels and loss functions are changing during training. This is critical for debugging multi-stream models during training (e.g., isolating unstable channels, finiding channels that perform worse). We can make this optional via a flag to reduce clutter. This was part of issue https://github.com/ecmwf/WeatherGenerator/issues/29

### Describe the solution you'd like

Loss is computed individually [here](https://github.com/ecmwf/WeatherGenerator/blob/65f9c85b7f34f1ac7adcc44d467e8bd3f4377442/src/weathergen/train/trainer.py#L437C49-L437C68) and averaged over channels to be stored in [`losses_all`](https://github.com/ecmwf/WeatherGenerator/blob/65f9c85b7f34f1ac7adcc44d467e8bd3f4377442/src/weathergen/train/trainer.py#L442C25-L442C54). 

The solution is to add another dimension to `losses_all` and save the individual losses there. As this array is used anyway for logging, minimal changes is required within the code. This change would fully backwards compatible.

### Additional context

_No response_

### Organisation

AWI

### PR DESCRIPTION
**Pull request by kacpnowak**
## Description

Implements logging losses per channel, now taking into account that ranks can have different number of samples. 

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes #282 

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [x] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [x] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by kacpnowak** [comment]
Closes https://github.com/ecmwf/WeatherGenerator/issues/282

**Comment by kacpnowak** [comment]
Fixes #282

File: src/weathergen/utils/distributed.py [hunk 1]
```diff
@@ -33,3 +36,79 @@ def is_root(pg: dist.ProcessGroup | None = None) -> bool:
...
+
+
+def get_rank() -> int:
+    """
+    Get current rank number
+
+    Returns:
+        int: current rank
+    """
+    if not dist.is_available():
```
Comment (tjhunter) [inline src/weathergen/utils/distributed.py:60]: _is_distributed_initialized

File: src/weathergen/utils/distributed.py [hunk 2]
```diff
@@ -33,3 +36,79 @@ def is_root(pg: dist.ProcessGroup | None = None) -> bool:
...
+        int: world size
+    """
+    if not dist.is_available():
+        return 1
+    if not dist.is_initialized():
+        return 1
+    return dist.get_world_size()
+
+
+def get_rank() -> int:
```
Comment (tjhunter) [inline src/weathergen/utils/distributed.py:53]: please update `is_root`. that function is following the best practices for pytorch. Maybe Seb Hoffman also has something to say about that part too.

File: src/weathergen/utils/distributed.py [hunk 3]
```diff
@@ -33,3 +36,79 @@ def is_root(pg: dist.ProcessGroup | None = None) -> bool:
...
+        int: current rank
+    """
+    if not dist.is_available():
+        return 0
+    if not dist.is_initialized():
+        return 0
+    return dist.get_rank()
+
+
+def all_gather(data: Tensor) -> list[Tensor]:
```
Comment (tjhunter) [inline src/weathergen/utils/distributed.py]: you should make very explicit that this implementation does _not_ allow gradient propagation (or does it? I would assume it breaks the tape tracking of the tensors but stranger things have happened in pytorch).

File: src/weathergen/train/trainer.py [hunk 4]
```diff
@@ -578,13 +590,44 @@ def train(self, epoch):
...
+            stddev_all (dict[str, torch.Tensor]): Dictionary mapping each stream name to its
+                per-channel standard deviation tensor.
+        """
+        losses_all: dict[str, Tensor] = {}
+        stddev_all: dict[str, Tensor] = {}
+
+        # Make list of losses into a tensor. This is individual tensor per rank
+        real_loss = torch.tensor(self.loss_model_hist, device=self.devices[0])
+        # Gather all tensors from all ranks into a list and stack them into one tensor again
+        real_loss = torch.cat(all_gather(real_loss))
```
Comment (tjhunter) [inline src/weathergen/train/trainer.py]: I am surprised it works as expected

**Comment by clessig** [comment]
I will try the code later today. But yes, would be great if others could test as well. Anyone from JSC?

________________________________

File: src/weathergen/utils/distributed.py (hunk 3 above)
Comment (kacpnowak) [reply src/weathergen/utils/distributed.py]: it's a great quesiton. I couldn't find anything explicit on this topic, but in my understanding it's not allowing gradients to flow. The gradient flag is preserved but once it's reconstructed it's detached from the autograd's graph

**Comment by kacpnowak** [comment]
Thanks for finding out this bug. I've patched it

**Review by clessig — COMMENTED** [review]
Everything looks good and is working but can we replace the `all_gather` with one that doesn't scratch on the bytes. This would do the job:

```
def all_gather_vdim(tensor: torch.Tensor, group=None) -> list[torch.Tensor]:
    """Gather tensors with different number of dimensions."""
    world_size = dist.get_world_size(group=group)
    # Gather shapes first
    shapes = all_gather_vlen(
        torch.as_tensor(tensor.shape, device=tensor.device), group=group
    )
    # Gather data
    inputs = [tensor] * world_size
    outputs = [
        torch.empty(*_shape, dtype=tensor.dtype, device=tensor.device)
        for _shape in shapes
    ]
    dist.all_to_all(outputs, inputs, group=group)
    return outputs
 ```
